### PR TITLE
Migrate some registry-replacements to vanilla class mixin injections

### PR DESCRIPTION
**Pull request by Alwinfy**
The reasons are twofold: 
- First, mixin reduces redundant code-- you can see that, for example, you have to VanillaCopy an entire function to change a single constant reference in ApothCactusBlock.java; mixin removes this requirement and makes updating easier (because you don't need to remember to re-copy the entire function).
- Second, mixins are generally more composable than RRs; if some other mod wanted to change an aspect of sugar cane (for example doubling the number of growths per tick, see: Snad), a mixin to a different site in randomTick wouldn't interfere with Apotheosis' change. However, with RR, only one mod's sugar-cane changes would take effect.

Of course, the changes are purely internal and the end-users get an identical experience:
Look ma, no changes!

### CODE REVIEW

**Review by ChloeDawn — CHANGES_REQUESTED** [review]
The method targets should use full descriptors for safer and more validatable targetting during compile across game versions. It is a practise recommended by Mumfrey and lack of doing so is only a habit introduced by the third-party IDEA plugin. Qualification of the target is not important, however.

File: build.gradle [hunk 1]
```diff
@@ -1,34 +1,47 @@
...
 
 archivesBaseName = name + '-' + mcVersion
 
 sourceCompatibility = targetCompatibility = compileJava.sourceCompatibility = compileJava.targetCompatibility = '1.8'
 
+mixin {
+    add sourceSets.main, 'apotheosis.mixins.refmap.json'
+}
+
+compileJava.options.compilerArgs << "-Xlint:all,-classfile,-processing,-deprecation" << "-Werror"
```
Comment (ChloeDawn) [inline build.gradle:25]: Why are you doing two left shifts here?

File: build.gradle [hunk 2]
```diff
@@ -67,6 +80,7 @@ jar {
                     "Implementation-Title": project.name,
                     "Implementation-Version": project.version,
                     "Implementation-Vendor" : project.author,
-                    "Implementation-Timestamp": new Date().format("yyyy-MM-dd'T'HH:mm:ssZ")],)
+                    "Implementation-Timestamp": new Date().format("yyyy-MM-dd'T'HH:mm:ssZ"),
+		    "MixinConfigs": "apotheosis.mixins.json",])
```
Comment (ChloeDawn) [inline build.gradle]: Mixed-use of tabs and spaces

File: build.gradle (hunk 1 above)
Comment (Alwinfy) [reply build.gradle:25]: different args, yo
think C++ streams lol

File: build.gradle [hunk 3]
```diff
@@ -1,34 +1,47 @@
...
 minecraft {
     mappings 'snapshot', project.properties.mappingsVersion
     accessTransformer = file('src/main/resources/META-INF/accesstransformer.cfg')
     runs {
         client = {
             properties 'forge.logging.markers': ''
             properties 'forge.logging.console.level': 'debug'
+            properties 'mixin.env.disableRefMap': 'true'
             workingDirectory project.file('run').canonicalPath
+            arg "-mixin.config=apotheosis.mixins.json"
```
Comment (ChloeDawn) [inline build.gradle]: Mixed-use of double-quote and single-quote strings

**Review by ChloeDawn — CHANGES_REQUESTED** [review]

File: build.gradle (hunk 1 above)
Comment (ChloeDawn) [reply build.gradle:25]: The args are just comma-separated, it could all be one string, no?

File: build.gradle (hunk 2 above)
Comment (Alwinfy) [reply build.gradle]: oops mb

File: src/main/java/shadows/apotheosis/mixin/MixinBambooBlock.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+@Mixin(BambooBlock.class)
+public class MixinBambooBlock {
+	@ModifyConstant(method = {
+		"randomTick",
+		"canGrow",
+		"grow(Lnet/minecraft/world/server/ServerWorld;Ljava/util/Random;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/BlockState;)V",
+		"getNumBambooBlocksAbove",
+		"getNumBambooBlocksBelow"
+	}, constant = @Constant(intValue = 16))
+	public int modifyMaxBambooHeight(int maxHeightIn) {
```
Comment (ChloeDawn) [inline src/main/java/shadows/apotheosis/mixin/MixinBambooBlock.java]: Mixin injectors should be **`private`**. `public` injectors cannot be obfuscated at runtime and run a significantly higher risk of conflict.

**Review by ChloeDawn — CHANGES_REQUESTED** [review]

File: src/main/resources/apotheosis.mixins.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  "compatibilityLevel": "JAVA_8",
+  "refmap": "apotheosis.mixins.refmap.json",
+  "package": "shadows.apotheosis.mixin",
+  "mixins": [
+    "MixinBambooBlock",
+    "MixinCactusBlock",
+    "MixinSugarCaneBlock"
+  ],
+  "injectors": {
+    "defaultRequire": 1
```
Comment (ChloeDawn) [inline src/main/resources/apotheosis.mixins.json]: In addition to this, mixin injectors should include the max number of injections they are expected to make in a target (see argument `allow`). Alternatively, this option can be omitted entirely and both `require` and `allow` declared explicitly on the injectors.

**Review by ChloeDawn — CHANGES_REQUESTED** [review]

File: src/main/resources/apotheosis.mixins.json [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  "refmap": "apotheosis.mixins.refmap.json",
+  "package": "shadows.apotheosis.mixin",
+  "mixins": [
+    "MixinBambooBlock",
+    "MixinCactusBlock",
+    "MixinSugarCaneBlock"
+  ],
+  "injectors": {
+    "defaultRequire": 1,
+    "defaultAllow": 1
```
Comment (ChloeDawn) [inline src/main/resources/apotheosis.mixins.json:14]: This option does not exist, unfortunately

**Review by ChloeDawn — CHANGES_REQUESTED** [review]

**Comment by Shadows-of-Fire** [comment]
I could write the reasons why I don't want to do this, but that's more time than I have

**Pull request closed without merging**